File: internal/publish/publish.go
package publish

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"strings"
)

// Feed pushes the current feed.xml to the configured git remote.
// No-op if there are no changes to commit.
func Feed(message string) error {
	if _, err := os.Stat("feed.xml"); err != nil {
		return fmt.Errorf("feed.xml not found; run fetch first: %w", err)
	}
	if message == "" {
		message = "Update feed.xml"
	}

	if err := run("git", "rev-parse", "--is-inside-work-tree"); err != nil {
		return fmt.Errorf("not a git repository")
	}

	status, err := output("git", "status", "--porcelain", "--", "feed.xml")
	if err != nil {
		return err
	}
	if strings.TrimSpace(status) == "" {
		// Still push in case a previous commit wasn't pushed.
		if err := run("git", "push"); err != nil {
			return fmt.Errorf("git push: %w", err)
		}
		return nil
	}

	if err := run("git", "add", "--", "feed.xml"); err != nil {
		return fmt.Errorf("git add: %w", err)
	}
	if err := run("git", "commit", "-m", message); err != nil {
		return fmt.Errorf("git commit: %w", err)
	}
	if err := run("git", "push"); err != nil {
		return fmt.Errorf("git push: %w", err)
	}
	return nil
}

func run(name string, args ...string) error {
	cmd := exec.Command(name, args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

func output(name string, args ...string) (string, error) {
	cmd := exec.Command(name, args...)
	var buf bytes.Buffer
	cmd.Stdout = &buf
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		return "", err
	}
	return buf.String(), nil
}

File: lib/publish.js
import { existsSync } from "node:fs";
import { spawnSync } from "node:child_process";

function run(name, args, { capture = false } = {}) {
  const result = spawnSync(name, args, {
    encoding: "utf8",
    stdio: capture ? ["ignore", "pipe", "inherit"] : "inherit",
  });
  if (result.error) throw result.error;
  if (result.status !== 0) {
    throw new Error(`${name} ${args.join(" ")} failed (exit ${result.status})`);
  }
  return capture ? result.stdout || "" : "";
}

/** @returns {string} Local timestamp as `yyyy/mm/dd HH:mm`. */
function stamp() {
  const d = new Date();
  const p = (n) => String(n).padStart(2, "0");
  return `${d.getFullYear()}/${p(d.getMonth() + 1)}/${p(d.getDate())} ${p(d.getHours())}:${p(d.getMinutes())}`;
}

/**
 * Commit and push feed XML files.
 * @param {string} message
 * @param {string[]} [files]
 */
export function publishFeed(message = `update all available feeds ${stamp()}`, files = ["facebook.xml", "twitter.xml", "feed.xml"]) {
  const existing = files.filter((f) => existsSync(f));
  if (!existing.length) {
    throw new Error("no feed xml files found; run fetch first");
  }

  try {
    run("git", ["rev-parse", "--is-inside-work-tree"], { capture: true });
  } catch {
    throw new Error("not a git repository");
  }

  const status = run("git", ["status", "--porcelain", "--", ...existing], {
    capture: true,
  }).trim();

  if (status) {
    run("git", ["add", "--", ...existing]);
    run("git", ["commit", "-m", message]);
  }

  run("git", ["push"]);
}
